CMake improvements

MaartenBent opened the pull request:
* Separate `BUILD_INTERFACE` and `INSTALL_INTERFACE` and install target files, see #2391, cc @JackBoosY 
* Use CMake's [target_precompile_headers](https://cmake.org/cmake/help/latest/command/target_precompile_headers.html) over `cotire` when available (CMake 3.16 or newer). Allow to specify `wxBUILD_PRECOMP=COTIRE` to force using `cotire`.
* Fixes for monolithic build.
* Fix some build warnings.

Commit:
CMake: Export target files

MaartenBent committed:
CMake: Set precompiled headers when adding libraries

Combine the three pch related functions into one.

MaartenBent committed:
CMake: Use target_precompile_headers when available

Keep using cotire for CMake versions older than 3.16, or when user specifies wxBUILD_PRECOMP=COTIRE.
The scintilla headers need to be specified for target_precompile_headers to work.
Enable it when the target has at least 2 source files (same as cotire does).

MaartenBent committed:
CMake: Fix monolithic build

Don't target net and webview libraries directly, but use the wx_ function that will add it to the mono library when applicable.

MaartenBent committed:
CMake: Fix building tests and sample after making target includes private

MaartenBent committed:
Fix unused variable warnings

vadz commented on samples/image/canvas.cpp:
```diff
@@ -386,6 +386,9 @@ MyCanvas::MyCanvas( wxWindow *parent, wxWindowID id,
     // my_png_from_mem = wxBitmap::NewFromPNGData(cursor_png, WXSIZEOF(cursor_png));
     my_png_from_mem = wxBITMAP_PNG_FROM_DATA(cursor);
+
+    // prevent -Wunused-const-variable when compiler fails to detect it usage
```
```suggestion
    // prevent -Wunused-const-variable when compiler fails to detect its usage
```

vadz commented on build/cmake/functions.cmake:
```diff
@@ -158,7 +160,7 @@ function(wx_set_common_target_properties target_name)
 
 # Set common properties on wx library target
-function(wx_set_target_properties target_name is_base)
+function(wx_set_target_properties target_name is_base is_plugin)
```
Could we perhaps use a single string, like `base`, `gui` or `plugin` instead of 2 booleans? Even though this is not C++, I still find calls to this function below passing it `FALSE TRUE` arguments perfectly unreadable...

MaartenBent replied:
Yes, I'll an argument for this.

vadz commented on build/cmake/functions.cmake:
```diff
@@ -297,13 +301,14 @@ function(wx_set_target_properties target_name is_base)
-            ${wxSOURCE_DIR}/include
+            $<BUILD_INTERFACE:${wxSETUP_HEADER_PATH}>
+            $<BUILD_INTERFACE:${wxSOURCE_DIR}/include>
+            $<INSTALL_INTERFACE:include>
```
Is `setup.h` installed under `include` by CMake? I'm surprised that we don't need to mention the `lib/wx` subdirectory here.

MaartenBent replied:
It probably is, I'l add it.

vadz commented on build/cmake/lib/CMakeLists.txt:
```diff
@@ -99,8 +103,10 @@ if(wxBUILD_MONOLITHIC)
+        add_dependencies(wxmono ${dep})
+    endforeach()
     wx_webview_copy_webview2_loader(wxmono)
-    wx_finalize_lib(wxmono)
```
Nice to not need it any more, thanks!

vadz reviewed (commented):
Thanks! I'm not qualified to review this, so I'll just merge it if nobody else says anything during the next couple of days.

MaartenBent committed:
CMake: Make all target_compile_definitions private

It is only used to build scintilla, no need to expose these definitions to external projects.

MaartenBent committed:
CMake: Make all target_include_directories private

There is no need to expose them to external projects.

MaartenBent committed:
CMake: Improve handling plugin in wx_set_target_properties

Don't link with toolkit or wx libraries, and set the correct defines.

MaartenBent committed:
CMake: Export WXUSINGDLL for shared builds

Always define WXUSINGDLL for plugins.

MaartenBent commented:
I'm looking through all the defines, since some of them seem unused. I already removed `_LIB`, and I think it is also safe to remove `DLL_EXPORTS`. I can't find any use of it in the source files.
It's probably there because the `.vcxproj` files define them and we tried to keep CMake the same. The `.vcxproj` files also have `_USRDLL` which seems unused too.
So I'll remove `DLL_EXPORTS` from CMake unless there is a reason to keep it?

vadz commented:
All of those are predefined in the standard MSVS projects, i.e. those created by the IDE. I don't think anybody can rely on them, however, as they should only affect wx itself and not any code using it, so it should be fine to remove them.

MaartenBent committed:
CMake: Remove duplicate and unused defines

Remove unused CMake code to get number of mono source files.
wxUSE_BASE=1 is also defined in wx_set_target_properties.
There seems no reason to define _LIB and DLL_EXPORTS, wx does not use it. They
were probably added because the vcxproj files have them.

MaartenBent commented:
I removed them now. CMake does add default defines to MSVS projects (like `_DEBUG` or `WIN32`) but not these ones.

I don't plan any more changes for this PR. So lets wait if there are any comments, and merge otherwise.

MaartenBent committed:
CMake: Apply source groups using a function

This will add source groups to all projects, not only libraries.
And there is no need anymore to include it in every project file.

MaartenBent committed:
CMake: Add source group for CMake files

Add all of the CMake related files together in one group/filter.
This includes the CMakeLists.txt of the project, cmake_pch.[hxx,cxx] when
using CMake's precompiled header feature, and a .rule file created by cotire.

Also include manifest files in the Resource group. Even though they are not
visible (in VS2022), they would create an empty MSW Headers group for samples.

talregev commented:
@JackBoosY Your suggestions here?

JackBoosY commented on build/cmake/functions.cmake:
```diff
@@ -15,7 +15,9 @@ if(CMAKE_GENERATOR STREQUAL "Xcode")
     set(COTIRE_UNITY_SOURCE_EXCLUDE_EXTENSIONS "" CACHE STRING "wxWidgets override of cotire exclude")
 endif()
-include(cotire)                        # For precompiled header handling
+if((wxBUILD_PRECOMP STREQUAL "ON" AND CMAKE_VERSION VERSION_LESS "3.16") OR (wxBUILD_PRECOMP STREQUAL "COTIRE"))
+    include(cotire) # For precompiled header handling
+endif()
```
I think we always need to deprecate cotire in the future, but not it's okay.

MaartenBent replied:
wxWidgets supports old distributions that come with old versions of CMake. Those should also be able to benefit from PCH, that is why I kept it.

JackBoosY commented on build/cmake/functions.cmake:
```diff
@@ -294,16 +300,19 @@ function(wx_set_target_properties target_name is_base)
             )
     endif()
 
+    file(RELATIVE_PATH wxSETUP_HEADER_REL ${wxOUTPUT_DIR} ${wxSETUP_HEADER_PATH})
```
What about debug build?

MaartenBent replied:
`wxSETUP_HEADER_PATH` contains a generator expression for debug build. The result in the target file looks like:
```
INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/lib/vc_x64_lib/mswu\$<\$<CONFIG:Debug>:d>;${_IMPORT_PREFIX}/include"
```

JackBoosY commented on build/cmake/functions.cmake:
```diff
@@ -294,16 +300,19 @@ function(wx_set_target_properties target_name is_base)
+    if(wxTOOLKIT_INCLUDE_DIRS AND NOT wxTARGET_IS_BASE)
         target_include_directories(${target_name}
-            PUBLIC ${wxTOOLKIT_INCLUDE_DIRS})
+            PRIVATE ${wxTOOLKIT_INCLUDE_DIRS})
```
I have no idea with this part. Other guys should review this part again.

MaartenBent replied:
There is no need to expose all the stuff used to build wxWidgets. Projects using it don't need to know that wxWidgets includes some gtk or libsecret or gstreamer or webkit headers. The only thing they need is the wxWidgets headers.

JackBoosY commented on build/cmake/functions.cmake:
```diff
@@ -342,22 +351,23 @@ function(wx_set_target_properties target_name is_base)
     if(wxBUILD_SHARED)
         string(TOUPPER ${target_name_short} target_name_upper)
         if(target_name_short STREQUAL "mono")
-            target_compile_definitions(${target_name} PRIVATE DLL_EXPORTS WXMAKINGDLL)
-        else()
-            target_compile_definitions(${target_name} PRIVATE DLL_EXPORTS WXMAKINGDLL_${target_name_upper})
```
Why we don't need `DLL_EXPORTS` anymore?

MaartenBent replied:
wxWidgets does not use it. And things like `__declspec(dllexport)` are covered via `WXMAKINGDLL_*`, see `wx/dlimpexp.h`.

JackBoosY commented on build/cmake/functions.cmake:
```diff
@@ -342,22 +351,23 @@ function(wx_set_target_properties target_name is_base)
-        endif()
-        if(NOT target_name_short STREQUAL "base")
-            target_compile_definitions(${target_name} PRIVATE WXUSINGDLL)
+            target_compile_definitions(${target_name} PRIVATE WXMAKINGDLL)
```
Maybe
```suggestion
            target_compile_definitions(${target_name} PRIVATE $<BUILD_INTERFACE:WXMAKINGDLL> PUBLIC $<INSTALL_INTERFACE:WXUSINGDLL>)
```

MaartenBent replied:
This `target_compile_definitions` is conditional to only the `mono` project, so it cannot be combined with the one below that applies to all projects.

JackBoosY commented on build/cmake/functions.cmake:
```diff
@@ -342,22 +351,23 @@ function(wx_set_target_properties target_name is_base)
+        elseif(NOT wxTARGET_IS_PLUGIN)
+            target_compile_definitions(${target_name} PRIVATE WXMAKINGDLL_${target_name_upper})
         endif()
+        target_compile_definitions(${target_name} INTERFACE WXUSINGDLL)
```
And remove this line.

JackBoosY commented on build/cmake/functions.cmake:
```diff
@@ -342,22 +351,23 @@ function(wx_set_target_properties target_name is_base)
+        target_compile_definitions(${target_name} INTERFACE WXUSINGDLL)
+    endif()
+    if(wxTARGET_IS_PLUGIN OR (wxBUILD_SHARED AND NOT target_name_short STREQUAL "base"))
+        target_compile_definitions(${target_name} PRIVATE WXUSINGDLL)
```
Same with above.

MaartenBent replied:
Same as the other comment. This define is conditional to some projects, while `INTERFACE WXUSINGDLL` is for all projects.

JackBoosY commented on build/cmake/functions.cmake:
```diff
@@ -422,42 +432,35 @@ macro(wx_add_library name)
             set(runtime_dir "bin")
         endif()
         wx_install(TARGETS ${name}
+            EXPORT ${name}Targets
```
I think the easier way is export all the targets in one group files such as `wxWidgetsTargets.cmake`, so we don't need to include all the targets files in `wxWidgetsConfig.cmake`.

MaartenBent replied:
Yes, I'll do this.

JackBoosY commented on build/cmake/functions.cmake:
```diff
@@ -422,42 +432,35 @@ macro(wx_add_library name)
             RUNTIME DESTINATION "${runtime_dir}/${GEN_EXPR_DIR_FIX}${wxPLATFORM_LIB_DIR}"
             BUNDLE DESTINATION Applications/wxWidgets
             )
+        wx_install(EXPORT ${name}Targets NAMESPACE wx:: DESTINATION "lib/${wxPLATFORM_LIB_DIR}/cmake")
```
So we can just call this line once.

JackBoosY commented on build/cmake/functions.cmake:
```diff
@@ -422,42 +432,35 @@ macro(wx_add_library name)
+            get_target_property(target_source_files ${target_name} SOURCES)
+            list(FILTER target_source_files INCLUDE REGEX ".*(.cpp|.cxx)$")
+            list(LENGTH target_source_files target_source_count)
+            if(target_source_count GREATER_EQUAL 2)
+                target_precompile_headers(${target_name} PRIVATE "$<$<COMPILE_LANGUAGE:CXX>:${prec_header}>")
+            endif()
```
`else`?

MaartenBent replied:
Building a PCH for a project with only one source files adds more overhead than is gained by the PCH. Cotire has a minimum of two files too.
I'll add a comment to clarify.

JackBoosY commented on build/cmake/functions.cmake:
```diff
@@ -579,7 +580,8 @@ function(wx_set_builtin_target_properties target_name)
     wx_set_common_target_properties(${target_name} DEFAULT_WARNINGS)
     if(NOT wxBUILD_SHARED)
-        wx_install(TARGETS ${name} ARCHIVE DESTINATION "lib/${GEN_EXPR_DIR_FIX}${wxPLATFORM_LIB_DIR}")
+        wx_install(TARGETS ${name} EXPORT ${name}Targets ARCHIVE DESTINATION "lib/${GEN_EXPR_DIR_FIX}${wxPLATFORM_LIB_DIR}")
+        wx_install(EXPORT ${name}Targets NAMESPACE wx:: DESTINATION "lib/${wxPLATFORM_LIB_DIR}/cmake")
```
Same with above.

JackBoosY commented on build/cmake/functions.cmake:
```diff
@@ -836,6 +835,7 @@ function(wx_add name group)
     elseif(group STREQUAL Tests)
         target_include_directories(${target_name} PRIVATE ${wxSOURCE_DIR}/tests)
         target_include_directories(${target_name} PRIVATE ${wxSOURCE_DIR}/3rdparty/catch/include)
+        target_include_directories(${target_name} PRIVATE ${wxTOOLKIT_INCLUDE_DIRS})
```
Also need other guys to review this: should we only use these folders for build, not for usage?

MaartenBent replied:
This is for `wxWidgets` own projects, like `test_base` and `test_gui`. It uses `glib` directly so it should also include those headers.

JackBoosY commented on build/cmake/lib/CMakeLists.txt:
```diff
@@ -85,22 +83,20 @@ if(wxBUILD_MONOLITHIC)
-            target_compile_definitions(wxmono ${vis} ${wxMONO_DEFINITIONS_${vis}})
-        endif()
     endforeach()
+    if(wxMONO_INCLUDE_DIRS)
+        target_include_directories(wxmono BEFORE PRIVATE ${wxMONO_INCLUDE_DIRS})
+    endif()
+    if(wxMONO_DEFINITIONS)
+        target_compile_definitions(wxmono PRIVATE ${wxMONO_DEFINITIONS})
+    endif()
```
I'm not sure about this part.

JackBoosY commented on build/cmake/lib/base/CMakeLists.txt:
```diff
@@ -26,28 +24,24 @@ endif()
-
 if(wxUSE_ZLIB)
-    wx_lib_include_directories(wxbase PRIVATE ${ZLIB_INCLUDE_DIRS})
+    wx_lib_include_directories(wxbase ${ZLIB_INCLUDE_DIRS})
```
In my mind, I think we should pass `PUBLIC` / `PRIVATE` / `INTERFACE` into `wx_lib_include_directories` so we can easily figure out which keyword we used here.

MaartenBent replied:
The point of this change is that all the headers that `wxWidgets` includes are `PRIVATE`. It doesn't have to be in the interface, because other projects don't need them. They only need the `wxWidgets` headers.

JackBoosY commented on build/cmake/lib/stc/CMakeLists.txt:
```diff
@@ -164,31 +162,21 @@ target_include_directories(wxscintilla PRIVATE
     ${wxSOURCE_DIR}/src/stc/scintilla/src
     )
-target_compile_definitions(wxscintilla PUBLIC
+target_compile_definitions(wxscintilla PRIVATE
```
Also need other guys to review this.

JackBoosY commented on build/cmake/lib/stc/CMakeLists.txt:
```diff
@@ -164,31 +162,21 @@ target_include_directories(wxscintilla PRIVATE
-    wx_target_enable_precomp(wxscintilla ${wxSCINTILLA_PREC_HEADER})
-endif()
-
+wx_target_enable_precomp(wxscintilla "${wxSOURCE_DIR}/src/stc/scintilla/include/Scintilla.h")
```
Why remove the condition?

MaartenBent replied:
The check for the condition is inside `wx_target_enable_precomp`.

MaartenBent reviewed (commented):
Thanks for your review and comments. I answered most of them.

MaartenBent committed:
Fix unused parameter warning in macOS build

MaartenBent committed:
CMake: Fix using precompiled headers with Objective-C++ files

Compiling .mm files gives error 'Objective C file was disabled in PCH but is currently enabled'.

Enable the OBJCXX language so the Objective-C++ compiler will be detected and used for .mm files instead of the c++ compiler.
Detect if there are .mm files and enable PCH for them too.

Don't set COTIRE_UNITY_SOURCE_EXCLUDE_EXTENSIONS when cotire is not used.

MaartenBent committed:
CMake: Add IS_MONO option to wx_add_library and wx_set_target_properties

Use this instead of checking for the target name.

MaartenBent committed:
CMake: Fix different identifier warning when building with XCode

Add plist file to the Resource group.

MaartenBent committed:
CMake: Create only one target file

Put the target file in lib/cmake/wxWidgets directory.

JackBoosY reviewed (approved):
Okay, LGTM for now.
